### PR TITLE
WIP: Canary

### PR DESCRIPTION
**Pull request by christianalfoni**
This is the implementation of RFC: https://github.com/cerebral/cerebral/issues/1184

```js
import { Controller, Module, Provider, Compute, CerebralError } from 'cerebral'
```

These are the constructs that Cerebral consists of and you use them like this:

### Module
The module is how you structure your application.

```js
const app = Module({
  state: {},
  signals: {},
  providers: {},
  modules: {},
  catch:[]
})

// Or callback
const app = Module(({name, path, controller}) => ({
  state: {}
})
```

1. Providers are now an object instead of an array. This makes a more sense as providers needs a name and we do not need this `provide` factory:

```js
const app = Module({
  state: {},
  providers: {
    http: HttpProvider({})
  }
})
```

2. The only way to `catch` an error now is at the module level. Errors propagate up to parent modules if not caught. Meaning there is no signal specific error catching or "global". "Global" would be your root level module.

```js
const app = Module({
  state: {},
  catch: [
    [HttpProviderError, sequences.catchHttpError]
  ]
})
```

### Controller
The controller takes a module as first argument, and other options as a second argument.

```js
const controller = Controller(rootModule, {
  devtools: null
})
```

### Provider
Providers will now have a stronger concept. The concept is being the API between your application and your tools of choice. It can only be defined in one way, an object with methods. This makes things consistent and also enforces the idea that you think of a provider as a "bridge" between your app and the tools your are using.

```js
const myProvider = Provider({
  foo() {}
})
```

You can point to the existing context using `this.context`. It works this way to simplify with just one API surface and we can do prototypal optimizations under the hood.
```js
const api = Provider({
  getUser() {
    return this.context.http.get('/user')
  }
})
```

All providers defined this way is automatically optimized and wrapped for debugger purposes. If you just want to use a 3rd party tool directly, you can still do that by just attaching it:

```js
Module({
  providers: { uuid }
})
```

### Compute

This is just a change to be consistent with the other APIs:

```js
const myComputed = Compute(state`foo`, (foo) => ())
```

### CerebralError
To more easily create different error types you can now define your errors by extending the CerebralError:

```js
import { CerebralError } from 'cerebral'

export class AppError extends CerebralError {}
export class ApiError extends CerebralError {}
```

### Module tag
You can now use a `module` tag to point to the state of the module running the signal:

```js
[
  set(module`foo`, 'bar')
]
```

### Migration

1. Change controller configuration to take in a top level module
2. Wrap all modules in Module constructor
3. Change configuration of providers to use object
4. Change any custom providers to use the Provider factory

### CODE REVIEW

File: docs/api/module.md [hunk 1]
```diff
@@ -1,40 +1,42 @@
 # Module
 
-Modules help you structure your state and signals into logical units. You can think of them as namespaces for state and signals.
+Modules are the core building blocks of your application. This is where you define state, signals, other modules etc.
```
Comment (henri-hulski) [inline docs/api/module.md]: Maybe better 'submodules' instead of 'other modules'.

File: docs/api/tags.md [hunk 2]
```diff
@@ -12,6 +12,7 @@ state`items.${props`itemKey`}`
 This allows you to express signals and component dependencies more effectively. There are 4 tags in Cerebral.
 
 - **state** - Used in signals and connect to target state
+- **module** - Used in signals and connect to target state of the module running the signal
```
Comment (henri-hulski) [inline docs/api/tags.md]: How do you use use 'module' in connect as the view is not bound to a specific module.

File: docs/api/operators.md [hunk 3]
```diff
@@ -5,13 +5,13 @@ Read more about operators in the [Cerebral in depth - Operators](https://www.jsb
 
 ## State operators
 
-The methods for changing state within actions is also available as operators. All state operators support using both **state** and **props** tags as values.
+The methods for changing state within actions is also available as operators. All state operators support using **state**, **module** and **props** tags as values.
```
Comment (henri-hulski) [inline docs/api/operators.md]: Should be "within actions **are** ...".

File: docs/introduction/providers.md [hunk 4]
```diff
@@ -26,62 +27,49 @@ The [devtools](/docs/api/devtools) also adds its own provider called **debugger*
...
+    http
+  }
 })
 ```
 
-When using the **provide** factory the provider will automatically be wrapped by the Cerebral devtools, to track its usage in the debugger.
+**It does not matter what module exposes the provider, all providers are global. That means any provider defined is available in any action in the application.**
 
-Now this provider is available to any action:
+You can use providers for pretty much anything, though typically it is to handle some kind of side effect. Examples of providers is [@cerebral/storage](https://github.com/cerebral/cerebral/tree/master/packages/node_modules/@cerebral/storage), [@cerebral/firebase](https://github.com/cerebral/cerebral/tree/master/packages/node_modules/@cerebral/firebase) and [@cerebral/http](https://github.com/cerebral/cerebral/tree/master/packages/node_modules/@cerebral/http).
```
Comment (henri-hulski) [inline docs/introduction/providers.md]: Should be "Examples of providers **are** ...".

File: docs/api/tags.md (hunk 2 above)
Comment (christianalfoni) [reply docs/api/tags.md]: @henri-hulski It is only for use in operators. should note that :)

File: docs/api/module.md (hunk 1 above)
Comment (christianalfoni) [reply docs/api/module.md]: agree, adjusting :)

File: packages/node_modules/@cerebral/mobx-state-tree/README.md [hunk 5]
```diff
@@ -44,48 +42,41 @@ import { types } from 'mobx-state-tree'
...
-  // on the module, though nested models can of course also have views. Note
-  // that views are only to GET state, not to change it. Meaning that in actions
-  // do not use views with for example state.set etc., only state.get
-  views(state) {
-    return {
-      currentBox() {
-        return state.boxes[state.currentBoxId]
-      }
+  // Define getters
+  // store.upper // "BAR"
```
Comment (henri-hulski) [inline packages/node_modules/@cerebral/mobx-state-tree/README.md]: @christianalfoni the method in the comment is different from the function name ("upper" -> "upperName").

File: packages/node_modules/@cerebral/mobx-state-tree/README.md [hunk 6]
```diff
@@ -44,48 +42,41 @@ import { types } from 'mobx-state-tree'
...
+  getters: {
+    upperName() {
+      return this.name.toUpperCase()
+    }
+  },
+  // Define computed (receives a value to compute)
+  // store.helloFromFoo("someName")
+  computed: {
+    helloFrom(name) {
+      return `Hello ${this.name}, I am ${name}`
```
Comment (henri-hulski) [inline packages/node_modules/@cerebral/mobx-state-tree/README.md]: Here also:
"helloFromFoo" -> "helloFrom"

File: docs/api/signal.md [hunk 7]
```diff
@@ -193,45 +173,24 @@ function myAction ({path}) {
...
 ```js
-function myAction ({someProvider, path}) {
+function myAction ({ someProvider, path }) {
   return someProvider.doAsync()
-    .then((result) => path.bananas({data: result.data}))
+    .then((result) => path.bananas({ data: result.data }))
 }
 ```
-
-## Errors
```
Comment (Guria) [inline docs/api/signal.md]: leave and add removal note?

File: docs/api/tags.md [hunk 8]
```diff
@@ -2,17 +2,24 @@
...
-import {state, props} from 'cerebral/tags'
+import { state, props } from 'cerebral/tags'
 
 state`items.${state`currentItemKey`}`
 
 state`items.${props`itemKey`}`
 ```
 
-This allows you to express signals and component dependencies more effectively. There are 4 tags in Cerebral.
+There are 6 tags in Cerebral.
```
Comment (Guria) [inline docs/api/tags.md]: Seems it's better not to count tags here :)

**Comment by Guria** [comment]
Changed to Approve, docs shouldn't block getting into `next`

File: packages/node_modules/@cerebral/mobx-state-tree/README.md (hunk 6 above)
Comment (christianalfoni) [reply packages/node_modules/@cerebral/mobx-state-tree/README.md]: 👍

File: docs/api/action.md [hunk 9]
```diff
@@ -122,12 +132,5 @@ function someActionFactory(someArgument) {
 You have access to the controller instance on the context:
 
 ```js
-function someAction ({controller}) {}
-```
-
-## Execution
```
[the remark this answers is not in the document]
Comment (christianalfoni) [reply docs/api/action.md]: I just removed docs as this is very low level. Do not see any reason to use it for application development.

File: docs/api/signal.md (hunk 7 above)
Comment (christianalfoni) [reply docs/api/signal.md]: We have it in migration docs, think that is enough :)

**Pull request merged**